File: PE/PEFormat/PEFileReader.ts
// <reference path="PEFile.ts" />
// <reference path="DosHeader.ts" />
// <reference path="PEHeader.ts" />
// <reference path="OptionalHeader.ts" />
// <reference path="SectionHeader.ts" />

// <reference path="MZSignature.ts" />
// <reference path="PESignature.ts" />
// <reference path="PEMagic.ts" />
// <reference path="DllCharacteristics.ts" />

// <reference path="SectionCharacteristics.ts" />

// <reference path="../IO/BinaryReader.ts" />

module Mi.PE.PEFormat {
    var dosHeaderSize: number = 64;

    export function readPEFile(peFile: PEFile, reader: Mi.PE.IO.BinaryReader) {
        if (!peFile.dosHeader)
            peFile.dosHeader = new DosHeader();

        readDosHeader(peFile.dosHeader, reader);

        if (peFile.dosHeader.lfanew > dosHeaderSize)
            peFile.dosStub = reader.readBytes(peFile.dosHeader.lfanew - dosHeaderSize);

        if (!peFile.peHeader)
            peFile.peHeader = new PEHeader();

        readPEHeader(peFile.peHeader, reader);

        if (!peFile.optionalHeader)
            peFile.optionalHeader = new OptionalHeader();

        readOptionalHeader(peFile.optionalHeader, reader);

        if (peFile.peHeader.numberOfSections > 0) {
            peFile.sectionHeaders = Array(peFile.peHeader.numberOfSections);
            for (var i = 0; i < peFile.sectionHeaders.length; i++) {
                var sectionHeader = new SectionHeader();
                readSectionHeader(sectionHeader, reader);
                reader.addSection(sectionHeader.rawData, sectionHeader.virtualRange);
                peFile.sectionHeaders[i] = sectionHeader;
            }
        }
    }

    function readDosHeader(dosHeader: DosHeader, reader: Mi.PE.IO.BinaryReader) {
        dosHeader.mz = reader.readShort();
        if (dosHeader.mz != MZSignature.MZ)
            throw new Error("MZ signature is invalid: " + (<number>(dosHeader.mz)).toString(16) + "h.");

        dosHeader.cblp = reader.readShort();
        dosHeader.cp = reader.readShort();
        dosHeader.crlc = reader.readShort();
        dosHeader.cparhdr = reader.readShort();
        dosHeader.minalloc = reader.readShort();
        dosHeader.maxalloc = reader.readShort();
        dosHeader.ss = reader.readShort();
        dosHeader.sp = reader.readShort();
        dosHeader.csum = reader.readShort();
        dosHeader.ip = reader.readShort();
        dosHeader.cs = reader.readShort();
        dosHeader.lfarlc = reader.readShort();
        dosHeader.ovno = reader.readShort();

        dosHeader.res1 =
            reader.readInt() +
            ((reader.readInt() << 14) * 4); // JavaScript can handle large integers, but cannot do '<<' around them.

        dosHeader.oemid = reader.readShort();
        dosHeader.oeminfo = reader.readShort();

        dosHeader.reserved =
        [
            reader.readInt(),
            reader.readInt(),
            reader.readInt(),
            reader.readInt(),
            reader.readInt()
        ];

        dosHeader.lfanew = reader.readInt();
    }

    function readPEHeader(peHeader: PEHeader, reader: Mi.PE.IO.BinaryReader) {
        peHeader.pe = reader.readInt();
        if (peHeader.pe != <number>PESignature.PE)
            throw new Error("PE signature is invalid: " + (<number>(peHeader.pe)).toString(16) + "h.");

        peHeader.machine = reader.readShort();
        peHeader.numberOfSections = reader.readShort();
        var timestampNum = reader.readInt();
        var timestamp = new Date(timestampNum * 1000);
        var timestamp = new Date(
            Date.UTC(
                timestamp.getFullYear(),
                timestamp.getMonth(),
                timestamp.getDate(),
                timestamp.getHours(),
                timestamp.getMinutes(),
                timestamp.getSeconds(),
                timestamp.getMilliseconds()));

        peHeader.timestamp = timestamp;

        peHeader.pointerToSymbolTable = reader.readInt();
        peHeader.numberOfSymbols = reader.readInt();
        peHeader.sizeOfOptionalHeader = reader.readShort();
        peHeader.characteristics = reader.readShort();
    }

    function readOptionalHeader(optionalHeader: OptionalHeader, reader: Mi.PE.IO.BinaryReader) {
        optionalHeader.peMagic = <PEMagic>reader.readShort();

        if (optionalHeader.peMagic != PEMagic.NT32
            && optionalHeader.peMagic != PEMagic.NT64)
            throw Error("Unsupported PE magic value " + (<number>optionalHeader.peMagic).toString(16) + "h.");

        optionalHeader.linkerVersion = new Version(reader.readByte(), reader.readByte());

        optionalHeader.sizeOfCode = reader.readInt();
        optionalHeader.sizeOfInitializedData = reader.readInt();
        optionalHeader.sizeOfUninitializedData = reader.readInt();
        optionalHeader.addressOfEntryPoint = reader.readInt();
        optionalHeader.baseOfCode = reader.readInt();

        if (optionalHeader.peMagic == PEMagic.NT32) {
            optionalHeader.baseOfData = reader.readInt();
            optionalHeader.imageBase = reader.readInt();
        }
        else {
            optionalHeader.imageBase = reader.readLong();
        }

        optionalHeader.sectionAlignment = reader.readInt();
        optionalHeader.fileAlignment = reader.readInt();
        optionalHeader.operatingSystemVersion = new Version(reader.readShort(), reader.readShort());
        optionalHeader.imageVersion = new Version(reader.readShort(), reader.readShort());
        optionalHeader.subsystemVersion = new Version(reader.readShort(), reader.readShort());
        optionalHeader.win32VersionValue = reader.readInt();
        optionalHeader.sizeOfImage = reader.readInt();
        optionalHeader.sizeOfHeaders = reader.readInt();
        optionalHeader.checkSum = reader.readInt();
        optionalHeader.subsystem = <Subsystem>reader.readShort();
        optionalHeader.dllCharacteristics = <DllCharacteristics>reader.readShort();

        if (optionalHeader.peMagic == PEMagic.NT32) {
            optionalHeader.sizeOfStackReserve = reader.readInt();
            optionalHeader.sizeOfStackCommit = reader.readInt();
            optionalHeader.sizeOfHeapReserve = reader.readInt();
            optionalHeader.sizeOfHeapCommit = reader.readInt();
        }
        else {
            optionalHeader.sizeOfStackReserve = reader.readLong();
            optionalHeader.sizeOfStackCommit = reader.readLong();
            optionalHeader.sizeOfHeapReserve = reader.readLong();
            optionalHeader.sizeOfHeapCommit = reader.readLong();
        }

        optionalHeader.loaderFlags = reader.readInt();
        optionalHeader.numberOfRvaAndSizes = reader.readInt();

        if (optionalHeader.dataDirectories == null
            || optionalHeader.dataDirectories.length != optionalHeader.numberOfRvaAndSizes)
            optionalHeader.dataDirectories = <DataDirectory[]>(Array(optionalHeader.numberOfRvaAndSizes));

        for (var i = 0; i < optionalHeader.dataDirectories.length; i++) {
            optionalHeader.dataDirectories[i] = readDataDirectory(reader);
        }
    }

    function readDataDirectory(reader: Mi.PE.IO.BinaryReader) {
        var virtualAddress = reader.readInt();
        var size = reader.readInt();
        return new DataDirectory(virtualAddress, size);
    }

    function readSectionHeader(sectionHeader: SectionHeader, reader: Mi.PE.IO.BinaryReader) {
        sectionHeader.name = reader.readZeroFilledAscii(8);

        var virtualSize = reader.readInt();
        var virtualAddress = reader.readInt();
        sectionHeader.virtualRange = new DataDirectory(virtualAddress, virtualSize);

        var sizeOfRawData = reader.readInt();
        var pointerToRawData = reader.readInt();
        sectionHeader.rawData = new DataDirectory(pointerToRawData, sizeOfRawData);

        sectionHeader.pointerToRelocations = reader.readInt();
        sectionHeader.pointerToLinenumbers = reader.readInt();
        sectionHeader.numberOfRelocations = reader.readShort();
        sectionHeader.numberOfLinenumbers = reader.readShort();
        sectionHeader.characteristics = <SectionCharacteristics>reader.readInt();
    }
}

File: Zoom.PE/Mi.PE/PEFile.Reader.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.IO;

namespace Mi.PE
{
    using Mi.PE.Internal;
    using Mi.PE.PEFormat;

    partial class PEFile
    {
        internal static readonly DateTime TimestampEpochUTC = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);

        public void ReadFrom(BinaryStreamReader reader)
        {
            ReadDosHeader(reader);

            reader.Position = this.DosHeader.lfanew;
            ReadPEHeader(reader);
            ReadOptionalHeader(reader);

            this.SectionHeaders = new SectionHeader[this.PEHeader.NumberOfSections];

            for (int i = 0; i < this.SectionHeaders.Length; i++)
            {
                var sectionHeader = new SectionHeader();
                ReadSectionHeader(reader, sectionHeader);
                this.SectionHeaders[i] = sectionHeader;
            }
        }

        void ReadDosHeader(BinaryStreamReader reader)
        {
            if (this.DosHeader == null)
                this.DosHeader = new DosHeader();

            var signature = (MZSignature)reader.ReadInt16();
            if (signature != MZSignature.MZ)
                throw new BadImageFormatException("MZ signature expected, " + ((ushort)signature).ToString("X4") + "h found.");

            this.DosHeader.cblp = reader.ReadUInt16();
            this.DosHeader.cp = reader.ReadUInt16();
            this.DosHeader.crlc = reader.ReadUInt16();
            this.DosHeader.cparhdr = reader.ReadUInt16();
            this.DosHeader.minalloc = reader.ReadUInt16();
            this.DosHeader.maxalloc = reader.ReadUInt16();
            this.DosHeader.ss = reader.ReadUInt16();
            this.DosHeader.sp = reader.ReadUInt16();
            this.DosHeader.csum = reader.ReadUInt16();
            this.DosHeader.ip = reader.ReadUInt16();
            this.DosHeader.cs = reader.ReadUInt16();
            this.DosHeader.lfarlc = reader.ReadUInt16();
            this.DosHeader.ovno = reader.ReadUInt16();

            this.DosHeader.res1 = reader.ReadUInt64();

            this.DosHeader.oemid = reader.ReadUInt16();
            this.DosHeader.oeminfo = reader.ReadUInt16();

            this.DosHeader.ReservedNumber0 = reader.ReadUInt32();
            this.DosHeader.ReservedNumber1 = reader.ReadUInt32();
            this.DosHeader.ReservedNumber2 = reader.ReadUInt32();
            this.DosHeader.ReservedNumber3 = reader.ReadUInt32();
            this.DosHeader.ReservedNumber4 = reader.ReadUInt32();
            this.DosHeader.lfanew = reader.ReadUInt32();

            if (this.DosHeader.lfanew > DosHeader.Size)
            {
                this.DosStub = new byte[this.DosHeader.lfanew - DosHeader.Size];
                reader.ReadBytes(this.DosStub, 0, this.DosStub.Length);
            }
        }

        void ReadPEHeader(BinaryStreamReader reader)
        {
            if (this.PEHeader == null)
                this.PEHeader = new PEHeader();

            var peSignature = (PESignature)reader.ReadUInt32();
            if (peSignature != PESignature.PE00)
                throw new BadImageFormatException("PE00 signature expected, " + ((ushort)peSignature).ToString("X8") + "h found.");

            this.PEHeader.Machine = (Machine)reader.ReadInt16();
            this.PEHeader.NumberOfSections = reader.ReadUInt16();
            uint timestampNum = reader.ReadUInt32();
            this.PEHeader.Timestamp = TimestampEpochUTC.AddSeconds(timestampNum);
            this.PEHeader.PointerToSymbolTable = reader.ReadUInt32();
            this.PEHeader.NumberOfSymbols = reader.ReadUInt32();
            this.PEHeader.SizeOfOptionalHeader = reader.ReadUInt16();
            this.PEHeader.Characteristics = (ImageCharacteristics)reader.ReadInt16();
        }

        void ReadOptionalHeader(BinaryStreamReader reader)
        {
            if (this.OptionalHeader == null)
                this.OptionalHeader = new OptionalHeader();

            var peMagic = (PEMagic)reader.ReadInt16();

            if (peMagic != PEMagic.NT32
                && peMagic != PEMagic.NT64)
                throw new BadImageFormatException("Unsupported PE magic value " + peMagic + ".");

            this.OptionalHeader.PEMagic = peMagic;

            this.OptionalHeader.MajorLinkerVersion = reader.ReadByte();
            this.OptionalHeader.MinorLinkerVersion = reader.ReadByte();
            this.OptionalHeader.SizeOfCode = reader.ReadUInt32();
            this.OptionalHeader.SizeOfInitializedData = reader.ReadUInt32();
            this.OptionalHeader.SizeOfUninitializedData = reader.ReadUInt32();
            this.OptionalHeader.AddressOfEntryPoint = reader.ReadUInt32();
            this.OptionalHeader.BaseOfCode = reader.ReadUInt32();

            if (peMagic == PEMagic.NT32)
            {
                this.OptionalHeader.BaseOfData = reader.ReadUInt32();
                this.OptionalHeader.ImageBase = reader.ReadUInt32();
            }
            else
            {
                this.OptionalHeader.ImageBase = reader.ReadUInt64();
            }

            this.OptionalHeader.SectionAlignment = reader.ReadUInt32();
            this.OptionalHeader.FileAlignment = reader.ReadUInt32();
            this.OptionalHeader.MajorOperatingSystemVersion = reader.ReadUInt16();
            this.OptionalHeader.MinorOperatingSystemVersion = reader.ReadUInt16();
            this.OptionalHeader.MajorImageVersion = reader.ReadUInt16();
            this.OptionalHeader.MinorImageVersion = reader.ReadUInt16();
            this.OptionalHeader.MajorSubsystemVersion = reader.ReadUInt16();
            this.OptionalHeader.MinorSubsystemVersion = reader.ReadUInt16();
            this.OptionalHeader.Win32VersionValue = reader.ReadUInt32();
            this.OptionalHeader.SizeOfImage = reader.ReadUInt32();
            this.OptionalHeader.SizeOfHeaders = reader.ReadUInt32();
            this.OptionalHeader.CheckSum = reader.ReadUInt32();
            this.OptionalHeader.Subsystem = (Subsystem)reader.ReadUInt16();
            this.OptionalHeader.DllCharacteristics = (DllCharacteristics)reader.ReadUInt16();

            if (peMagic == PEMagic.NT32)
            {
                this.OptionalHeader.SizeOfStackReserve = reader.ReadUInt32();
                this.OptionalHeader.SizeOfStackCommit = reader.ReadUInt32();
                this.OptionalHeader.SizeOfHeapReserve = reader.ReadUInt32();
                this.OptionalHeader.SizeOfHeapCommit = reader.ReadUInt32();
            }
            else
            {
                this.OptionalHeader.SizeOfStackReserve = reader.ReadUInt64();
                this.OptionalHeader.SizeOfStackCommit = reader.ReadUInt64();
                this.OptionalHeader.SizeOfHeapReserve = reader.ReadUInt64();
                this.OptionalHeader.SizeOfHeapCommit = reader.ReadUInt64();
            }

            this.OptionalHeader.LoaderFlags = reader.ReadUInt32();
            var rvaCountHeaderOffsetBase = reader.Position;
            this.OptionalHeader.NumberOfRvaAndSizes = reader.ReadUInt32();

            if (this.OptionalHeader.DataDirectories == null
                || this.OptionalHeader.DataDirectories.Length != this.OptionalHeader.NumberOfRvaAndSizes)
                this.OptionalHeader.DataDirectories = new DataDirectory[this.OptionalHeader.NumberOfRvaAndSizes];

            for (int i = 0; i < this.OptionalHeader.DataDirectories.Length; i++)
            {
                if (i == (int)DataDirectoryKind.Clr)
                {
                    var clrDirHeaderOffsetDelta = (reader.Position - rvaCountHeaderOffsetBase) / 4;
                }

                var dd = new DataDirectory();
                dd.Read(reader);
                this.OptionalHeader.DataDirectories[i] = dd;
            }
        }

        static void ReadSectionHeader(BinaryStreamReader reader, SectionHeader section)
        {
            section.Name = reader.ReadFixedZeroFilledAsciiString(SectionHeader.MaximumNameSize);
            section.VirtualSize = reader.ReadUInt32();
            section.VirtualAddress = reader.ReadUInt32();
            section.SizeOfRawData = reader.ReadUInt32();
            section.PointerToRawData = reader.ReadUInt32();
            section.PointerToRelocations = reader.ReadUInt32();
            section.PointerToLinenumbers = reader.ReadUInt32();
            section.NumberOfRelocations = reader.ReadUInt16();
            section.NumberOfLinenumbers = reader.ReadUInt16();
            section.Characteristics = (SectionCharacteristics)reader.ReadUInt32();
        }
    }
}